README.md
<p align="center">
  <img width="400" src="assets/v2.png" alt="we-validator" />
</p>
<p align="center">
  <a href="https://travis-ci.org/ChanceYu/we-validator"><img src="https://img.shields.io/travis/ChanceYu/we-validator/master.svg?color=green&style=popout-square" /></a>
  <a href="https://opensource.org/licenses/mit-license.php"><img src="https://img.shields.io/npm/l/we-validator.svg?color=blue&style=popout-square" /></a>
  <a href="https://www.npmjs.com/package/we-validator"><img src="https://img.shields.io/npm/v/we-validator.svg?color=green&style=popout-square" /></a>
</p>

> 2.x 和 1.x 版本有所差别，如果您继续使用老版本，可以查看 [1.x 版本](https://github.com/ChanceYu/we-validator/tree/v1)。如果您要升级到 2.x 请先阅读[升级指南](./UPGRADE.md)，推荐使用 2.x 版本。

简单灵活的表单验证插件，支持小程序、浏览器、Nodejs。小程序端支持：微信、支付宝、百度智能、字节跳动，小程序默认提示使用 `showToast`。

[API 文档](#api) | [示例 Examples](#examples)，如果你 :sparkling_heart: 就点击 **★Star** 吧。

## 特点

- 使用简单灵活，不依赖任何框架
- 既支持原生小程序方式，也支持 mpvue、wepy、taro 等小程序框架使用
- 支持浏览器以及 Nodejs 端使用
- [支持自定义规则](#wevalidatoraddrulerulename-ruleoption)
- [支持自定义错误信息提示](#wevalidatoronmessage)
- [支持动态添加或移除字段校验](#addrulesoptions)
- [支持单独校验一个或多个字段规则](#checkfieldsdata-fields-onmessage)
- [支持函数校验](#wevalidatorcheckvaluerulename-value-param)
- [支持多个字段同时校验并显示错误](#多个字段同时校验并显示错误)
- [默认支持常用校验规则](#默认支持的规则)



## 安装
使用 npm:

```bash
npm install we-validator --save
```

使用 cdn:

```html
<script src="https://unpkg.com/we-validator/dist/we-validator.min.js"></script>
```


## 使用
下面是微信小程序的用法，其它小程序类似

<details open>
<summary>点击查看栗子 :chestnut: </summary>

```html
<form bindsubmit="onSubmitForm">
    <input type="text" name="username" placeholder="用户名" />
    <input type="number" name="phoneno" placeholder="手机号" />
    <input type="text" name="str" placeholder="长度为3的字符串" />

    <button type="default" formType="submit">提交</button>
</form>
```

```javascript
const WeValidator = require('we-validator')

Page({
    onReady(){
        this.initValidator()
    },
    onSubmitForm(e){
        let { value } = e.detail

        if(!this.validatorInstance.checkData(value)) return

        // 开始提交表单
        // wx.request
    },
    initValidator(){
        // 实例化
        this.validatorInstance = new WeValidator({
            rules: {
                username: {
                    required: true
                },
                phoneno: {
                    required: true,
                    mobile: true
                },
                str: {
                    length: 3
                },
            },
            messages: {
                username: {
                    required: '请输入用户名'
                },
                phoneno: {
                    required: '请输入手机号',
                    mobile: '手机号格式不正确'
                },
                str: { // 非必填字段
                    length: '请输入长度为3的字符串'
                },
            },
        })
    },
})
```

</details>

## Examples
您也可参考当前项目下对应示例

- [原生微信小程序使用方式](./example/wechat/pages/index/index.js)
- [mpvue 使用方式](./example/mpvue/src/pages/index/index.vue)
- [web 浏览器使用方式](./example/web/index.html)
- [Nodejs 端使用方式](./example/nodejs/server/router/home.js)
- [复杂的校验案例-自定义规则-动态添加字段校验](./example/complex/index.html)


## API
 - [new WeValidator(options)](#new-wevalidatoroptions)
    - [.checkData(data, onMessage)](#checkdatadata-onmessage) 校验数据，会显示错误信息
    - [.checkFields(data, fields, onMessage)](#checkfieldsdata-fields-onmessage) 校验数据，会显示错误信息，只校验对应的字段
    - [.isValid(data, fields)](#isvaliddata-fields) 校验数据，不会显示错误信息，可单独校验特定字段
    - [.addRules(options)](#addrulesoptions) 动态添加字段校验
    - [.removeRules(fields)](#removerulesfields) 动态移除字段校验


## Static API
 - [WeValidator](#static-api)
    - [.addRule(ruleName, ruleOption)](#wevalidatoraddrulerulename-ruleoption) 添加自定义规则
    - [.checkValue(ruleName, value, param)](#wevalidatorcheckvaluerulename-value-param) 函数校验，单独校验某个内容
    - [.onMessage](#wevalidatoronmessage) 设置全局错误信息显示


## 默认支持的规则

> 注意: 非必填字段只有字段有值才会校验配置的规则

| 规则 | 描述 | 默认提示 |
| --- | --- | --- |
| `required: true` | 必填 | 此字段必填 |
| `pattern: /^\d+$/` | 正则通用 | 不符合此验证规则 |
| `email: true` | 电子邮件格式 | 请输入有效的电子邮件地址 |
| `mobile: true` | 11位手机号 | 请输入11位的手机号码 |
| `tel: true` | 座机号<br>例如：010-1234567、0551-1234567 | 请输入座机号 |
| `url: true` | URL网址 | 请输入有效的网址 |
| `idcard: true` | 身份证号 | 请输入18位的有效身份证 |
| `equalTo: 'field'` | 字段值相同校验<br>例如：密码和确认密码，[参考](./example/web/index.html#L82) | 输入值必须和字段 `field` 相同 |
| `notEqualTo: 'field'` | 字段值不能相同校验<br>与 `equalTo` 相反 | 输入值不能和字段 `field` 相同 |
| `contains: 'str'` | 是否包含某字符 | 输入值必须包含 `str` |
| `notContains: 'str'` | 不能包含某字符 | 输入值不能包含 `str` |
| `length: 5` | 长度为多少的字符串 | 请输入 `5` 个字符 |
| `minlength: 2` | 最少多长的字符串 | 最少要输入 `2` 个字符 |
| `maxlength: 6` | 最多多长的字符串 | 最多可以输入 `6` 个字符 |
| `rangelength: [2, 6]` | 某个范围长度的字符串 | 请输入长度在 `2` 到 `6` 之间的字符 |
| `number: true` | 数字 | 请输入有效的数字 |
| `digits: true` | 正整数数字 | 只能输入正整数数字 |
| `integer: true` | 正整数或负整数数字 | 只能输入整数数字 |
| `min: 3` | 大于多少的数字<br>（最小只能多少），也可以比较字段的值，[参考](./example/complex/index.html#L180) | 请输入大于 `3` 的数字 |
| `max: 9` | 小于多少的数字<br>（最大只能多少），也可以比较字段的值 | 请输入小于 `9` 的数字 |
| `range: [3, 9]` | 大于且小于多少的数字 | 请输入大于 `3` 且小于 `9` 的数字 |
| `chinese: true` | 中文字符 | 只能输入中文字符 |
| `minChinese: 3` | 最少多少个中文字符 | 最少输入 `3` 个中文字符 |
| `maxChinese: 9` | 最多多少个中文字符 | 最多输入 `9` 个中文字符 |
| `rangeChinese: [3, 9]` | 大于且小于多少个中文字符 | 只能输入 `3` 到 `9` 个中文字符 |
| `date: true` | 日期（默认使用 `new Date(value)` 校验） | 请输入有效的日期 |
| `dateISO: true` | 日期（ISO标准格式）<br>例如：2019-09-19，2019/09/19 | 请输入有效的日期（ISO标准格式） |
| `ipv4: true` | ipv4地址 | 请输入有效的IPv4地址 |

CHANGELOG.md
# Changelog

> 更新日志（推荐使用最新版本）

## Released

### [v2.1.17](https://github.com/ChanceYu/we-validator/releases/tag/v2.1.17)
- 修复：multiple 类型声明错误[issue](https://github.com/ChanceYu/we-validator/issues/25)


### [v2.1.16](https://github.com/ChanceYu/we-validator/releases/tag/v2.1.16)
- 修复：错误 message 的显示优化[pull](https://github.com/ChanceYu/we-validator/pull/16)
- 修复：修改空指针和语法修改[pull](https://github.com/ChanceYu/we-validator/pull/17)
- 修复：required: false 不进行“必须”验证[pull](https://github.com/ChanceYu/we-validator/pull/18)


### [v2.1.15](https://github.com/ChanceYu/we-validator/releases/tag/v2.1.15)
- 修复：部分ts定义错误[issue](https://github.com/ChanceYu/we-validator/issues/15)


### [v2.1.14](https://github.com/ChanceYu/we-validator/releases/tag/v2.1.14)
- 新增：添加ts的定义文件[merge](https://github.com/ChanceYu/we-validator/pull/14)


### [v2.1.11](https://github.com/ChanceYu/we-validator/releases/tag/v2.1.11)
- 修复：手机号以 `16` 开头的校验问题[issue](https://github.com/ChanceYu/we-validator/issues/12)


### [v2.1.10](https://github.com/ChanceYu/we-validator/releases/tag/v2.1.10)
- 新增：`.checkFields` 方法，可校验对应的字段，[参考](./README.md#checkfieldsdata-fields-onmessage)
- 修改：`.isValid` 方法，可校验对应的字段，[参考](./README.md#isvaliddata-fields)


### [v2.1.9](https://github.com/ChanceYu/we-validator/releases/tag/v2.1.9)
- 优化：部分规则
- 补充文档和增加动态字段校验使用示例


### [v2.1.8](https://github.com/ChanceYu/we-validator/releases/tag/v2.1.8)
- 修复：中文校验规则
- 修复：多个实例参数覆盖问题


### [v2.1.7](https://github.com/ChanceYu/we-validator/releases/tag/v2.1.7)
- 优化：校验逻辑


### [v2.1.6](https://github.com/ChanceYu/we-validator/releases/tag/v2.1.6)
- 修复：`WeValidator.checkValue` 校验空字段不符合预期
- 优化：`multiCheck` 为 `true` 时错误显示


### [v2.1.5](https://github.com/ChanceYu/we-validator/releases/tag/v2.1.5)
- 修改：通用正则校验
```diff
- regex
+ pattern
```
- 增加：1.x 到 2.x [升级指南](./UPGRADE.md)


### [v2.1.4](https://github.com/ChanceYu/we-validator/releases/tag/v2.1.4)
- 优化：自定义规则添加方式


### [v2.1.2](https://github.com/ChanceYu/we-validator/releases/tag/v2.1.2)
- 修改：静态方法 `WeValidator.checkValue`
```diff
- WeValidator.checkField
+ WeValidator.checkValue
```
- 修改文档


### [v2.1.1](https://github.com/ChanceYu/we-validator/releases/tag/v2.1.1)
- 新增：规则：`notEqualTo、notContains、integer、ipv4、ipv6`
- 新增：示例：[复杂的校验案例-自定义规则](./example/complex/index.html)


### [v2.1.0](https://github.com/ChanceYu/we-validator/releases/tag/v2.1.0)
- 重构：核心方法
- 重写：核心校验规则和默认提供的自定义规则
- 单独校验字段方法修改


### [v1](https://github.com/ChanceYu/we-validator/releases/tag/v1)（v1.0.0 - v1.4.3）
- v1 版本 `we-validator`

example/wechat/js/we-validator.js
/*!
 * we-validator
 * version: 2.1.17
 * address: https://github.com/ChanceYu/we-validator#readme
 * author:  ChanceYu <i.fish@foxmail.com>
 * license: MIT
 */
(function webpackUniversalModuleDefinition(root, factory) {
	if(typeof exports === 'object' && typeof module === 'object')
		module.exports = factory();
	else if(typeof define === 'function' && define.amd)
		define("WeValidator", [], factory);
	else if(typeof exports === 'object')
		exports["WeValidator"] = factory();
	else
		root["WeValidator"] = factory();
})(typeof self !== 'undefined' ? self : this, function() {
return /******/ (function(modules) { // webpackBootstrap
/******/ 	// The module cache
/******/ 	var installedModules = {};
/******/
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/
/******/ 		// Check if module is in cache
/******/ 		if(installedModules[moduleId]) {
/******/ 			return installedModules[moduleId].exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = installedModules[moduleId] = {
/******/ 			i: moduleId,
/******/ 			l: false,
/******/ 			exports: {}
/******/ 		};
/******/
/******/ 		// Execute the module function
/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
/******/
/******/ 		// Flag the module as loaded
/******/ 		module.l = true;
/******/
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/
/******/
/******/ 	// expose the modules object (__webpack_modules__)
/******/ 	__webpack_require__.m = modules;
/******/
/******/ 	// expose the module cache
/******/ 	__webpack_require__.c = installedModules;
/******/
/******/ 	// define getter function for harmony exports
/******/ 	__webpack_require__.d = function(exports, name, getter) {
/******/ 		if(!__webpack_require__.o(exports, name)) {
/******/ 			Object.defineProperty(exports, name, {
/******/ 				configurable: false,
/******/ 				enumerable: true,
/******/ 				get: getter
/******/ 			});
/******/ 		}
/******/ 	};
/******/
/******/ 	// getDefaultExport function for compatibility with non-harmony modules
/******/ 	__webpack_require__.n = function(module) {
/******/ 		var getter = module && module.__esModule ?
/******/ 			function getDefault() { return module['default']; } :
/******/ 			function getModuleExports() { return module; };
/******/ 		__webpack_require__.d(getter, 'a', getter);
/******/ 		return getter;
/******/ 	};
/******/
/******/ 	// Object.prototype.hasOwnProperty.call
/******/ 	__webpack_require__.o = function(object, property) { return Object.prototype.hasOwnProperty.call(object, property); };
/******/
/******/ 	// __webpack_public_path__
/******/ 	__webpack_require__.p = "";
/******/
/******/ 	// Load entry module and return exports
/******/ 	return __webpack_require__(__webpack_require__.s = 0);
/******/ })
/************************************************************************/
/******/ ([
/* 0 */
/***/ (function(module, exports, __webpack_require__) {

"use strict";


var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _rules = __webpack_require__(1);

var _rules2 = _interopRequireDefault(_rules);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var requiredFn = _rules2.default.required.rule;

/**
 * 环境检测
 */
var isWx = typeof wx !== 'undefined' && !!wx.showToast; // 微信小程序
var isMy = typeof my !== 'undefined' && !!my.showToast; // 支付宝小程序
var isSwan = typeof swan !== 'undefined' && !!swan.showToast; // 百度智能小程序
var isTt = typeof tt !== 'undefined' && !!tt.showToast; // 字节跳动小程序
var isBrowser = typeof window !== 'undefined' && !!window.alert; // 普通浏览器

var objString = Object.prototype.toString;

var isArray = Array.isArray || function (v) {
  return objString.call(v) === '[object Array]';
};
var isFunction = function isFunction(v) {
  return objString.call(v) === '[object Function]';
};
var isRegExp = function isRegExp(v) {
  return objString.call(v) === '[object RegExp]';
};

var WeValidator = function () {

  /**
   * 默认参数
   * @param {object} options
   * @param {object} [options.rules] 验证字段的规则
   * @param {object} [options.messages] 验证字段错误的提示信息
   * @param {function} [options.onMessage] 错误信息显示方式
   * @param {boolean} [options.multiCheck] 是否同时校验多个字段
   */
  function WeValidator() {
    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    _classCallCheck(this, WeValidator);

    this.options = options;

    this.required = requiredFn;
    this._checkAllRules();
  }

  /**
   * 所有校验规则
   */


  /**
   * 动态添加验证规则
   * @param {string} ruleName 规则名称
   * @param {object} ruleOption 规则配置
   * @param {string} [ruleOption.message] 默认错误信息文字
   * @param {regexp|function} [ruleOption.rule] 验证规则
   */


  /**
   * 验证单个字段数据
   * @param {string} ruleName 规则名称
   * @param {string} value 要验证的值
   * @param {any} param 传递的验证参数
   * @param {boolean} skip 未填跳过校验，仅供内部使用
   */


  _createClass(WeValidator, [{
    key: '_showErrorMessage',


    /**
     * 显示错误信息
     * @param {object} params 错误信息
     * @param {function} onMessage 自定义提示函数
     */
    value: function _showErrorMessage(params, onMessage) {
      // validatorInstance.checkData(data, onMessage)
      if (isFunction(onMessage)) {
        return onMessage(params);
      }

      // 参数形式 new WeValidator({ onMessage })
      if (isFunction(this.options.onMessage)) {
        return this.options.onMessage(params);
      }

      // 全局配置 WeValidator.onMessage
      if (isFunction(WeValidator.onMessage)) {
        return WeValidator.onMessage(params);
      }

      // 微信小程序
      if (isWx) {
        return wx.showToast({
          title: params.msg,
          icon: 'none'
        });
      }

      // 支付宝小程序
      if (isMy) {
        return my.showToast({
          content: params.msg,
          type: 'none'
        });
      }

      // 百度小程序
      if (isSwan) {
        return swan.showToast({
          title: params.msg,
          icon: 'none'
        });
      }

      // 字节跳动小程序
      if (isTt) {
        return tt.showToast({
          title: params.msg,
          icon: 'none'
        });
      }

      // 浏览器端
      if (isBrowser) alert(params.msg);
    }

    /**
     * 获取错误信息内容
     * @param {string} ruleName 规则名称
     * @param {string} attr 字段名称
     * @param {any} param 规则参数
     */

  }, {
    key: '_getErrorMessage',
    value: function _getErrorMessage(ruleName, attr, param) {
      var messages = this.options.messages;
      var defaultMessage = WeValidator.RULES[ruleName].message;

      if (messages && messages.hasOwnProperty(attr) && messages[attr][ruleName]) {
        defaultMessage = messages[attr][ruleName];
      }

      if (defaultMessage) {
        defaultMessage = defaultMessage.replace(/\{(\d)\}/g, function ($0, $1) {
          if (isArray(param)) {
            return param[$1];
          } else {
            return param;
          }
        });

        return defaultMessage;
      }
    }

    /**
     * 验证配置规则是否无效
     * @param {string} ruleName 规则名称
     * @param {string} attr 字段名称
     */

  }, {
    key: '_isRuleInvalid',
    value: function _isRuleInvalid(ruleName, attr) {
      if (!WeValidator.RULES.hasOwnProperty(ruleName)) {
        console.warn && console.warn('\u6CA1\u6709\u6B64\u9A8C\u8BC1\u89C4\u5219\uFF1A' + ruleName + '\uFF0C\u5B57\u6BB5\uFF1A' + attr);
        return true;
      }
    }

    /**
     * 验证所有配置规则是否正确
     */

  }, {
    key: '_checkAllRules',
    value: function _checkAllRules() {
      var _rules_ = this.options.rules;

      // 遍历字段
      for (var attr in _rules_) {
        // 遍历验证规则
        for (var ruleName in _rules_[attr]) {
          if (this._isRuleInvalid(ruleName, attr)) continue;
        }
      }
    }

    /**
     * 校验数据，会验证所有配置的字段规则
     * @param {object} data 验证的数据对象
     * @param {function} onMessage 自定义错误信息提示
     * @param {boolean} showMessage 是否显示提示信息，默认显示（内部使用）
     * @param {object} fieldMap 校验的字段，默认校验所有字段（内部使用）
     */

  }, {
    key: 'checkData',
    value: function checkData(data, onMessage) {
      var showMessage = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
      var fieldMap = arguments[3];

      var _rules_ = this.options.rules;
      var multiCheck = this.options.multiCheck;
      var hasError = false;
      var errorData = {};

      this.data = data;

      // 遍历字段
      for (var attr in _rules_) {
        if (fieldMap && !fieldMap.hasOwnProperty(attr)) continue;

        // 遍历验证规则
        for (var ruleName in _rules_[attr]) {
          if (this._isRuleInvalid(ruleName, attr)) continue;

          if (fieldMap) {
            var res = fieldMap[attr];
            if (isArray(res) && res.indexOf(ruleName) === -1) continue;
          }

          var ruleParam = _rules_[attr][ruleName];
          var value = '';

          if (data.hasOwnProperty(attr)) {
            value = data[attr];
          }

          if (isFunction(ruleParam)) {
            ruleParam = ruleParam.call(this, value);
          }

          var isFieldValid = WeValidator.checkValue.call(this, ruleName, value, ruleParam, true);

          // 验证不通过
          if (!isFieldValid) {
            hasError = true;

            var msg = this._getErrorMessage(ruleName, attr, ruleParam);
            var errorParam = null;

            if (showMessage && msg) {
              errorParam = {
                name: attr,
                value: value,
                param: ruleParam,
                rule: ruleName,
                msg: msg
              };
              errorData[attr] = errorParam;
            }

            if (!multiCheck) {
              if (errorParam) {
                this._showErrorMessage(errorParam, onMessage);
              }
              return false;
            }
          }
        }
      }

      if (hasError) {
        if (multiCheck && showMessage) {
          this._showErrorMessage(errorData, onMessage);
        }
        return false;
      }

      return true;
    }

    /**
     * 校验数据，只校验对应的字段规则
     * @param {object} data 验证的数据对象
     * @param {array} fields 校验的字段
     * @param {function} onMessage 自定义错误信息提示
     * @param {boolean} showMessage 是否显示提示信息，默认显示（内部使用）
     */

  }, {
    key: 'checkFields',
    value: function checkFields(data, fields, onMessage) {
      var showMessage = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : true;

      if (!isArray(fields)) throw new Error('第二个参数须为数组');

      // fields: [ '[field]:[rule]' ]
      // fields: [ 'phoneNo' ]  =>  { phoneNo: true }
      // fields: [ 'phoneNo:required' ]  =>  { phoneNo: ['required'] }
      // fields: [ 'phoneNo:required,mobile' ]  =>  { phoneNo: ['required', 'mobile'] }
      var fieldMap = {};

      fields.forEach(function (item) {
        var arr = item.split(':');
        var field = arr[0];
        var rules = arr[1];

        if (rules) {
          // 只校验特定规则
          rules = rules.split(',');
          fieldMap[field] = rules;
        } else {
          // 校验 field 字段的所有规则
          fieldMap[field] = true;
        }
      });

      return this.checkData(data, onMessage, showMessage, fieldMap);
    }

    /**
     * 校验数据，不会提示错误信息
     * @param {object} data 验证的数据对象
     * @param {array} fields 校验的字段。如果有，只校验对应的字段规则，默认校验所有配置的字段规则
     */

  }, {
    key: 'isValid',
    value: function isValid(data, fields) {
      if (isArray(fields)) {
        return this.checkFields(data, fields, null, false);
      } else {
        return this.checkData(data, null, false);
      }
    }

    /**
     * 动态添加字段校验
     * @param {object} options 配置参数
     * @param {object} [options.rules] 规则
     * @param {object} [options.messages] 提示消息
     */

  }, {
    key: 'addRules',
    value: function addRules() {
      var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      Object.assign(this.options.rules, options.rules || {});
      Object.assign(this.options.messages, options.messages || {});

      this._checkAllRules();
    }

    /**
     * 动态移除字段校验
     * @param {array} fields 要删除校验的字段
     */

  }, {
    key: 'removeRules',
    value: function removeRules(fields) {
      if (!isArray(fields)) throw new Error('参数须为数组');

      for (var i = 0; i < fields.length; i++) {
        var key = fields[i];

        delete this.options.rules[key];
      }
    }
  }]);

  return WeValidator;
}();

WeValidator.RULES = {};

WeValidator.addRule = function (ruleName, ruleOption) {
  WeValidator.RULES[ruleName] = ruleOption;
};

WeValidator.checkValue = function (ruleName, value, param, skip) {
  var rule = WeValidator.RULES[ruleName].rule;

  if (isRegExp(rule)) {
    if (skip) {
      return !requiredFn(value) || rule.test(value);
    } else {
      return rule.test(value);
    }
  }

  if (isFunction(rule)) {
    if (ruleName === 'required') {
      return param && requiredFn(value);
    } else {
      if (skip) {
        return !requiredFn(value) || rule.call(this, value, param);
      } else {
        return rule.call(this, value, param);
      }
    }
  }
};

WeValidator.RULES = _rules2.default;
WeValidator.required = requiredFn;

module.exports = WeValidator;

/***/ }),
/* 1 */
/***/ (function(module, exports, __webpack_require__) {

"use strict";


module.exports = {
  /**
   * 必填
   */
  required: {
    message: '此字段必填',
    rule: function rule(value) {
      if (typeof value === 'number') {
        value = value.toString();
      } else if (typeof value === 'boolean') {
        return true;
      }
      return !!(value && value.length > 0);
    }
  },
  /**
   * 正则通用
   */
  pattern: {
    message: '不符合此验证规则',
    rule: function rule(value, param) {
      return param.test(value);
    }
  },
  /**
   * 电子邮件
   */
  email: {
    message: '请输入有效的电子邮件地址',
    rule: /^[a-zA-Z0-9.!#$%&'*+\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/
  },
  /**
   * 手机号码
   */
  mobile: {
    message: '请输入 11 位的手机号码',
    rule: /^1[3456789]\d{9}$/
  },
  /**
   * 座机号，例如：010-1234567、0551-1234567
   */
  tel: {
    message: '请输入座机号',
    rule: /^(\d{3,4}-)?\d{7,8}$/
  },
  /**
   * URL网址
   */
  url: {
    message: '请输入有效的网址',
    rule: /^(?:(?:(?:https?|ftp):)?\/\/)(?:\S+(?::\S*)?@)?(?:(?!(?:10|127)(?:\.\d{1,3}){3})(?!(?:169\.254|192\.168)(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)(?:\.(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)*(?:\.(?:[a-z\u00a1-\uffff]{2,})).?)(?::\d{2,5})?(?:[/?#]\S*)?$/i
  },
  /**
   * 身份证号
   */
  idcard: {
    message: '请输入 18 位的有效身份证',
    rule: /^[1-9]\d{5}[1-9]\d{3}((0\d)|(1[0-2]))(([0|1|2]\d)|3[0-1])\d{3}([0-9]|X)$/
  },
  /**
   * 字段值相同校验（例如：密码和确认密码）
   */
  equalTo: {
    message: '输入值必须和字段 {0} 相同',
    rule: function rule(value, param) {
      return value === this.data[param];
    }
  },
  /**
   * 字段值不相同校验，与 equalTo 相反
   */
  notEqualTo: {
    message: '输入值不能和字段 {0} 相同',
    rule: function rule(value, param) {
      return value !== this.data[param];
    }
  },
  /**
   * 是否包含某字符
   */
  contains: {
    message: '输入值必须包含 {0}',
    rule: function rule(value, param) {
      return value.indexOf(param) > -1;
    }
  },
  /**
   * 不能包含某字符
   */
  notContains: {
    message: '输入值不能包含 {0}',
    rule: function rule(value, param) {
      return value.indexOf(param) === -1;
    }
  },
  /**
   * 长度为多少的字符串
   */
  length: {
    message: '请输入 {0} 个字符',
    rule: function rule(value, param) {
      return value.length == param;
    }
  },
  /**
   * 最少多长的字符串
   */
  minlength: {
    message: '最少要输入 {0} 个字符',
    rule: function rule(value, param) {
      return value.length >= param;
    }
  },
  /**
   * 最多多长的字符串
   */
  maxlength: {
    message: '最多可以输入 {0} 个字符',
    rule: function rule(value, param) {
      return value.length <= param;
    }
  },
  /**
   * 某个范围长度的字符串
   */
  rangelength: {
    message: '请输入长度在 {0} 到 {1} 之间的字符',
    rule: function rule(value, param) {
      return value.length >= param[0] && value.length <= param[1];
    }
  },
  /**
   * 数字
   */
  number: {
    message: '请输入有效的数字',
    rule: /^(?:-?\d+|-?\d{1,3}(?:,\d{3})+)?(?:\.\d+)?$/
  },
  /**
   * 正整数数字
   */
  digits: {
    message: '只能输入正整数数字',
    rule: /^\d+$/
  },
  /**
   * 正整数或负整数数字
   */
  integer: {
    message: '只能输入整数数字',
    rule: /^-?\d+$/
  },
  /**
   * 大于多少的数字/字段值
   */
  min: {
    message: '请输入大于 {0} 的数字',
    rule: function rule(value, param) {
      if (typeof param === 'string') param = this.data[param];

      return value >= param;
    }
  },
  /**
   * 小于多少的数字/字段值
   */
  max: {
    message: '请输入小于 {0} 的数字',
    rule: function rule(value, param) {
      if (typeof param === 'string') param = this.data[param];

      return value <= param;
    }
  },
  /**
   * 大于且小于多少的数字
   */
  range: {
    message: '请输入大于 {0} 且小于 {1} 的数字',
    rule: function rule(value, param) {
      return value >= param[0] && value <= param[1];
    }
  },
  /**
   * 中文字符
   */
  chinese: {
    message: '只能输入中文字符',
    rule: /^[\u4e00-\u9fa5]+$/
  },
  /**
   * 最少多少个中文字符
   */
  minChinese: {
    message: '最少输入 {0} 个中文字符',
    rule: function rule(value, param) {
      return new RegExp('^[\u4E00-\u9FA5]{' + param + ',}$').test(value);
    }
  },
  /**
   * 最多多少个中文字符
   */
  maxChinese: {
    message: '最多输入 {0} 个中文字符',
    rule: function rule(value, param) {
      return new RegExp('^[\u4E00-\u9FA5]{1,' + param + '}$').test(value);
    }
  },
  /**
   * 大于且小于多少个中文字符
   */
  rangeChinese: {
    message: '只能输入 {0} 到 {1} 个中文字符',
    rule: function rule(value, param) {
      return new RegExp('^[\u4E00-\u9FA5]{' + param[0] + ',' + param[1] + '}$').test(value);
    }
  },
  /**
   * 日期
   */
  date: {
    message: '请输入有效的日期',
    rule: function rule(value) {
      return !/Invalid|NaN/.test(new Date(value).toString());
    }
  },
  /**
   * 日期（ISO标准格式）例如：2019-09-19，2019/09/19
   */
  dateISO: {
    message: '请输入有效的日期（ISO 标准格式）',
    rule: /^\d{4}[\/\-](0?[1-9]|1[012])[\/\-](0?[1-9]|[12][0-9]|3[01])$/
  },
  /**
   * ipv4地址
   */
  ipv4: {
    message: '请输入有效的 IPv4 地址',
    rule: /^(25[0-5]|2[0-4]\d|[01]?\d\d?)\.(25[0-5]|2[0-4]\d|[01]?\d\d?)\.(25[0-5]|2[0-4]\d|[01]?\d\d?)\.(25[0-5]|2[0-4]\d|[01]?\d\d?)$/i
  },
  /**
   * ipv6地址
   */
  ipv6: {
    message: '请输入有效的 IPv6 地址',
    rule: /^((([0-9A-Fa-f]{1,4}:){7}[0-9A-Fa-f]{1,4})|(([0-9A-Fa-f]{1,4}:){6}:[0-9A-Fa-f]{1,4})|(([0-9A-Fa-f]{1,4}:){5}:([0-9A-Fa-f]{1,4}:)?[0-9A-Fa-f]{1,4})|(([0-9A-Fa-f]{1,4}:){4}:([0-9A-Fa-f]{1,4}:){0,2}[0-9A-Fa-f]{1,4})|(([0-9A-Fa-f]{1,4}:){3}:([0-9A-Fa-f]{1,4}:){0,3}[0-9A-Fa-f]{1,4})|(([0-9A-Fa-f]{1,4}:){2}:([0-9A-Fa-f]{1,4}:){0,4}[0-9A-Fa-f]{1,4})|(([0-9A-Fa-f]{1,4}:){6}((\b((25[0-5])|(1\d{2})|(2[0-4]\d)|(\d{1,2}))\b)\.){3}(\b((25[0-5])|(1\d{2})|(2[0-4]\d)|(\d{1,2}))\b))|(([0-9A-Fa-f]{1,4}:){0,5}:((\b((25[0-5])|(1\d{2})|(2[0-4]\d)|(\d{1,2}))\b)\.){3}(\b((25[0-5])|(1\d{2})|(2[0-4]\d)|(\d{1,2}))\b))|(::([0-9A-Fa-f]{1,4}:){0,5}((\b((25[0-5])|(1\d{2})|(2[0-4]\d)|(\d{1,2}))\b)\.){3}(\b((25[0-5])|(1\d{2})|(2[0-4]\d)|(\d{1,2}))\b))|([0-9A-Fa-f]{1,4}::([0-9A-Fa-f]{1,4}:){0,5}[0-9A-Fa-f]{1,4})|(::([0-9A-Fa-f]{1,4}:){0,6}[0-9A-Fa-f]{1,4})|(([0-9A-Fa-f]{1,4}:){1,7}:))$/i
  }
};

/***/ })
/******/ ]);
});